Add more details about Kubelet -> Node plugin communication (#10)
pohly/enhancements: keps/sig-node/3063-dynamic-resource-allocation/README.md

This RPC is called by kubelet when a Pod that wants to use the
specified resource is scheduled on a node.  The Plugin SHALL assume
that this RPC will be executed on the node where the resource will be
used.  ResourceClaim.meta.Namespace, ResourceClaim.meta.UID,

> ResourceClaim.Name, ResourceClaim.meta.Namespace and

Reviewer: @bart0sh @pohly Sorry for seeing it after it's already merged, but there is "ResourceClaim.meta.Namespace" two times.

ResourceClaimStatus.AllocationResult should be passed to the Plugin
as parameters to identify the claim and perform resource preparation.

ResourceClaim parameters(namespace, UUID, name) are useful for debugging
and enable the resource driver to retrieve the full ResourceClaim object,
should that ever be needed (normally it shouldn't).

> The Plugin SHALL return device name and kind for allocated device[s].

Reviewer 2: ```suggestion
The Plugin SHALL return fully qualified device name[s].
```

This RPC is called by kubelet when a Pod that wants to use the
specified resource is scheduled on a node.  The Plugin SHALL assume
that this RPC will be executed on the node where the resource will be

> used.  Resource Claim UID and allocation result attributes should be

Reviewer 2: On chat, we discussed that claim name/namespace/UID should be passed to identify the claim. They are useful for debugging and enable the resource driver to retrieve the full object, should that ever be needed (normally it shouldn't).

As it stands now, the KEP talks about a "Resource Claim UID" without defining what that UID is.

Author: Thanks for pointing out to this. Fixed.

will depend on how resources are used (for example, how often new Pods are
created), therefore it is impossible to predict what reasonable SLOs might be.

> There shouldn't be a significant performance impact on the Kubelet side as
> Resource manager component will do its work similarly to the Volume manager:

Reviewer 2: Arguing that it will have little impact because it is similar to some other component seems a bit optimistic to me :sweat_smile: 

That other component might have significant impact, and then so will we. I would rephrase this:
```suggestion
Resource manager component will do its work similarly to the Volume manager, but the overhead and complexity should be lower:
```

Author: The similarity to Volume manager is not the only argument. It's just a start of the explanation. It simply refers to the fact that both upper level components are run as goroutines, similarly to Volume manager.

Reviewer 2: But it is not a valid argument... similarity doesn't imply anything regarding performance impact.

Author: It does in my opinion. If it would do it sequentially(i.e. just calling some functions, not a goroutines) it can slow down the Kubelet.

Reviewer 2: That it uses goroutines is a valid argument.

Let's just simplify the statements, then it will become clearer, okay? Start by stating that the resource manager will be similar to the volume manager. Describe how it will be simpler. Then conclude with an estimate of the the performance overhead.

Reviewer 2: Also beware that "performance" could mean many things. If we consider the time required to start a pod, then the resource manager is on the critical path and running its code in goroutines doesn't change anything.

Reviewer 2: Claiming "no significant performance impact" without being able to provide benchmark results that back up such a claim is risky. Reviewers will jump on this and ask for proof.

Author: >  Reviewers will jump on this and ask for proof.

Can you suggest how to put it better?

both upper level components of the Resource manager (reconciler and
desired state of world populator) will run as goroutines. Resource
preparation should be fairly quick as in most cases it simply
creates CDI file 1-3 Kb in size. Unpreparing resource usually means

> deleting CDI file, so it should be quick as well.

Reviewer 2: Add:

```suggestion
deleting CDI file, so it should be quick as well.
The complexity is lower than in the volume manager
because there is only one global operation needed (prepare vs. 
attach + publish for each pod) and reconstruction
after a kubelet restart is simpler (call NodePrepareResource
again vs. trying to determine whether volumes
are mounted).
```

will depend on how resources are used (for example, how often new Pods are
created), therefore it is impossible to predict what reasonable SLOs might be.

> Resource manager component will do its work similarly to the Volume manager, but the overhead and complexity should be lower:
> both upper level components of the Resource manager (reconciler and
> desired state of world populator) will run as goroutines. Resource
> preparation should be fairly quick as in most cases it simply
> creates CDI file 1-3 Kb in size. Unpreparing resource usually means
> deleting CDI file, so it should be quick as well.
> The complexity is lower than in the volume manager
> because there is only one global operation needed (prepare vs. 
> attach + publish for each pod) and reconstruction
> after a kubelet restart is simpler (call NodePrepareResource
> again vs. trying to determine whether volumes
> are mounted).

Reviewer 2: ```suggestion
The resource manager component will do its work similarly to the 
existing volume manager, but the overhead and complexity should
be lower:

* Resource preparation should be fairly quick as in most cases it simply
  creates CDI file 1-3 Kb in size. Unpreparing resource usually means
  deleting CDI file, so it should be quick as well.

* The complexity is lower than in the volume manager
  because there is only one global operation needed (prepare vs. 
  attach + publish for each pod).
  
* Reconstruction after a kubelet restart is simpler (call 
  NodePrepareResource again vs. trying to determine whether 
  volumes are mounted).
```

will depend on how resources are used (for example, how often new Pods are
created), therefore it is impossible to predict what reasonable SLOs might be.

The resource manager component will do its work similarly to the 
existing volume manager, but the overhead and complexity should
be lower:

* Resource preparation should be fairly quick as in most cases it simply
  creates CDI file 1-3 Kb in size. Unpreparing resource usually means

>   deleting CDI file, so it should be quick as well.

Reviewer 2: Beware that this ignores that "real" resource drivers may have to do more work to update resource tracking. We don't need to highlight this here, but it may come up in the KEP review.

Author: I understand that. However, we stated in the KEP that recommended way is to do all heavy work on allocation stage. We can point reviewers to that, I believe.

Extending the Device Plugins API to use [Container Device Interface](https://github.com/container-orchestrated-devices/container-device-interface)
would help address some of the requirements, but not all of them.

> NodePrepareResource and NodeUnprepareResource could be added as optional
> calls to the Device Plugins API. However, this would mean that

Reviewer 2: Let's be clearer about what "optional calls" means:

```suggestion
NodePrepareResource and NodeUnprepareResource could be added to the Device Plugins API and only get called for
resource claims.

However, this would mean that
```

developers of the Device Plugins would have to implement mandatory
API calls (ListAndWatch, Allocate), which could create confusion
as those calls are meaningless for the Dynamic Resource Allocation

> purposes.

Reviewer 2: ```suggestion
purposes.

Even worse, existing device plugins would have to implement the new
calls with stubs that return errors because the generated Go interface
will require them.
```
